components/Buttons/src/MDCFloatingButton.m

```diff
@@ -72,67 +105,406 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  // (A) The same spacing on either side of the label.
+  // (SP) The spacing between the image and title
+  // (CEI) Content Edge Insets
+  //
+  // The diagram above assumes an LTR user interface orientation
+  // and a .leadingIcon imageLocation for this button.
+
+  BOOL isLeadingIcon = self.imageLocation == MDCFloatingButtonImageLocationLeading;
+  UIEdgeInsets adjustedLayoutInsets =
+      isLeadingIcon ? internalLayoutSpacingInsets
+                    : MDFInsetsFlippedHorizontally(internalLayoutSpacingInsets);
+
+  const CGRect insetBounds = UIEdgeInsetsInsetRect(UIEdgeInsetsInsetRect(self.bounds,
+                                                                         adjustedLayoutInsets),
+                                                   self.contentEdgeInsets);
+
+  const CGFloat imageViewWidth = CGRectGetWidth(self.imageView.bounds);
+  const CGFloat boundsCenterY = CGRectGetMidY(insetBounds);
+  CGFloat titleWidthAvailable = CGRectGetWidth(insetBounds);
+  titleWidthAvailable -= imageViewWidth;
+  titleWidthAvailable -= self.imageTitleSpacing;
+
+  const CGFloat availableHeight = CGRectGetHeight(insetBounds);
+  CGSize titleIntrinsicSize
+      = [self.titleLabel sizeThatFits:CGSizeMake(titleWidthAvailable, availableHeight)];
+
+  const CGSize titleSize = CGSizeMake(MAX(0, MIN(titleIntrinsicSize.width, titleWidthAvailable)),
+                                      MAX(0, MIN(titleIntrinsicSize.height, availableHeight)));
+
+  CGPoint titleCenter;
+  CGPoint imageCenter;
+  BOOL isLTR =
+      self.mdf_effectiveUserInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionLeftToRight;
+
+  // If we are LTR with a leading image, the image goes on the left.
+  // If we are RTL with a trailing image, the image goes on the left.
+  if ((isLTR && isLeadingIcon) || (!isLTR && !isLeadingIcon)) {
+    const CGFloat imageCenterX = CGRectGetMinX(insetBounds) + (imageViewWidth / 2);
+    const CGFloat titleCenterX = CGRectGetMaxX(insetBounds) - (titleWidthAvailable / 2);
+    titleCenter = CGPointMake(titleCenterX, boundsCenterY);
+    imageCenter = CGPointMake(imageCenterX, boundsCenterY);
+  }
+  // If we are LTR with a trailing image, the image goes on the right.
+  // If we are RTL with a leading image, the image goes on the right.
+  else {
+    const CGFloat imageCenterX = CGRectGetMaxX(insetBounds) - (imageViewWidth / 2);
+    const CGFloat titleCenterX = CGRectGetMinX(insetBounds) + (titleWidthAvailable / 2);
+    imageCenter = CGPointMake(imageCenterX, boundsCenterY);
+    titleCenter = CGPointMake(titleCenterX, boundsCenterY);
+  }
+
+  self.imageView.center = imageCenter;
+  self.imageView.frame = UIEdgeInsetsInsetRect(self.imageView.frame, self.imageEdgeInsets);
+  self.titleLabel.center = titleCenter;
+  CGRect newBounds = CGRectStandardize(self.titleLabel.bounds);
+  self.titleLabel.bounds = (CGRect){newBounds.origin, titleSize};
+  self.titleLabel.frame = UIEdgeInsetsInsetRect(self.titleLabel.frame, self.titleEdgeInsets);
```

Comment:
Im worried that multiple calls to layoutsubviews will cause the title to keep shrinking. 2 lines above we use the existing bounds and then shrink it here.

Reply:
lgtm. I didn't see the center getting set 4 lines above.

Reply:
I don't think there's a risk of calling layout multiple times that the values will change, but I will add a test.  The title's center is computed each time based on the button bounds. The title's size is based on the `intrinsicContentSize` of the label and the available space for the title.  

I think we should also not manipulate the `frame` property on the button, since it may be transformed for the collapse/expand animations. Opened #2569 to track an improvement.